### PR TITLE
Add tour support to map

### PR DESCRIPTION
**Pull request by sam33rdhakal**
Updates:
* adds tour carousel
* adds display mode to map view (CurrentFloor and Tour)
         - dispalys tour stops when map view is in `Tour Mode`
         - displays everything is map when map is in `Current Floor` mode
* fixes the marker's layout  (base of marker was not traingular before )

### CODE REVIEW

File: base/src/main/res/values/dimens.xml [hunk 1]
```diff
@@ -38,4 +38,5 @@
     <dimen name="bottomAudioControlIconSize">32dp</dimen>
     <dimen name="large_icon_size">42dp</dimen>
     <dimen name="mapObjectDetailsViewHeight">135dp</dimen>
+    <dimen name="EightySix">86dp</dimen>
```
[the remark this answers is not in the document]
Comment (sam33rdhakal) [reply base/src/main/res/values/dimens.xml]: Updated.

File: db/src/main/kotlin/edu/artic/db/models/ArticAudioFile.kt [hunk 2]
```diff
@@ -101,6 +104,7 @@ fun ArticAudioFile.Translation.asAudioFileModel(): AudioFileModel {
  * handle that?
  */
 data class AudioFileModel(
+        val nid: String,
```
[the remark this answers is not in the document]
Comment (sam33rdhakal) [reply db/src/main/kotlin/edu/artic/db/models/ArticAudioFile.kt]: Done.

File: build.gradle [hunk 3]
```diff
@@ -57,7 +57,8 @@ buildscript {
                 exo_player_core         : "com.google.android.exoplayer:exoplayer-core:$exo_player_version",
                 exo_player_dash         : "com.google.android.exoplayer:exoplayer-dash:$exo_player_version",
                 exo_player_ui           : "com.google.android.exoplayer:exoplayer-ui:$exo_player_version",
-                exo_player_media_session: "com.google.android.exoplayer:extension-mediasession:$exo_player_version"
+                exo_player_media_session: "com.google.android.exoplayer:extension-mediasession:$exo_player_version",
+                view_pager_indicator    : "com.github.fuzz-productions:CutoutViewIndicator:v0.8.2"
```
[the remark this answers is not in the document]
Comment (sam33rdhakal) [reply build.gradle]: Updated.

File: map/src/main/kotlin/edu/artic/map/MapFragment.kt [hunk 4]
```diff
@@ -98,80 +118,10 @@ class MapFragment : BaseViewModelFragment<MapViewModel>() {
...
-                            "        \"visibility\": \"off\"\n" +
-                            "      }\n" +
-                            "    ]\n" +
-                            "  }\n" +
-                            "]")
-            )
+            map.setMapStyle(MapStyleOptions(requireActivity().assets.fileAsString(filename = "google_map_config.json")))
             map.setMinZoomPreference(17f)
             map.setMaxZoomPreference(22f)
+            map.setPadding(0, requireActivity().getStatusBarHeight(), 0, 0)
```
[the remark this answers is not in the document]
Comment (sam33rdhakal) [reply map/src/main/kotlin/edu/artic/map/MapFragment.kt]: Added.

File: map/src/main/kotlin/edu/artic/map/MapFragment.kt [hunk 5]
```diff
@@ -589,21 +616,57 @@ class MapFragment : BaseViewModelFragment<MapViewModel>() {
...
                 .into(object : SimpleTarget<Bitmap>() {
                     override fun onResourceReady(resource: Bitmap, transition: Transition<in Bitmap>?) {
-                        if (viewModel.currentZoomLevel === MapZoomLevel.Three && viewModel.currentFloor == floor) {
+                        /**
+                         * If map context is Tour, get the order number of the stop.
+                         */
+                        if (viewModel.currentFloor == floor || mapContext is MapViewModel.MapContext.Tour) {
+
+                            /**
+                             * If the tour is not in the current floor make the ui transparent
```
[the remark this answers is not in the document]
Comment (sam33rdhakal) [reply map/src/main/kotlin/edu/artic/map/MapFragment.kt]: Updated.

File: map/src/main/kotlin/edu/artic/map/MapFragment.kt [hunk 6]
```diff
@@ -624,6 +687,41 @@ class MapFragment : BaseViewModelFragment<MapViewModel>() {
...
+     */
+    private fun loadTourObject(articTour: ArticTour, floor: Int, mapContext: MapViewModel.MapContext) {
+        Glide.with(this)
+                .asBitmap()
+                .load(articTour.thumbnailFullPath)
+                .into(object : SimpleTarget<Bitmap>() {
+                    override fun onResourceReady(resource: Bitmap, transition: Transition<in Bitmap>?) {
+                        if (mapContext is MapViewModel.MapContext.Tour) {
+
+                            val markerAlpha = if (viewModel.currentFloor == floor) {
```
[the remark this answers is not in the document]
Comment (sam33rdhakal) [reply map/src/main/kotlin/edu/artic/map/MapFragment.kt]: Extracted into function.

File: map/src/main/kotlin/edu/artic/map/MapViewModel.kt [hunk 7]
```diff
@@ -34,13 +36,31 @@ import javax.inject.Inject
...
+
+    /**
+     * Class used to represent the context of map.
+     * Map view can be used with following different contexts
+     * <ul>
+     *     <li>General: Everything is displayed for selected floor </li>
+     *     <li>Tour   : Only the tour tourStopViewModels are displayed in the map.</li>
+     * </ul>
+     */
+    sealed class MapContext {
```
[the remark this answers is not in the document]
Comment (sam33rdhakal) [reply map/src/main/kotlin/edu/artic/map/MapViewModel.kt]: `DisplayMode` aligns more with the intent of that class. Thanks @Cliabhach

File: map/src/main/kotlin/edu/artic/map/MapViewModel.kt [hunk 8]
```diff
@@ -34,13 +36,31 @@ import javax.inject.Inject
...
+        class General() : MapContext()
+        class Tour(val tour: ArticTour) : MapContext()
```
[the remark this answers is not in the document]
Comment (sam33rdhakal) [reply map/src/main/kotlin/edu/artic/map/MapViewModel.kt]: Done.

File: base/src/main/java/edu/artic/base/utils/ActivityExtensions.kt [hunk 9]
```diff
@@ -73,3 +73,13 @@ fun Context.isResourceConstrained() : Boolean {
 
     return retVal
 }
+
+
+fun Activity.getStatusBarHeight(): Int {
+    var statusBarHeight = 0
+    val resourceId = resources.getIdentifier("status_bar_height", "dimen", "android")
+    if (resourceId > 0) {
```
[the remark this answers is not in the document]
Comment (sam33rdhakal) [reply base/src/main/java/edu/artic/base/utils/ActivityExtensions.kt]: Nice catch @Cliabhach 👍

File: map/src/main/kotlin/edu/artic/map/MapFragment.kt [hunk 10]
```diff
@@ -79,9 +83,19 @@ class MapFragment : BaseViewModelFragment<MapViewModel>() {
     private lateinit var baseGroundOverlay: GroundOverlay
     private lateinit var buildingGroundOverlay: GroundOverlay
     private var groundOverlayGenerated: Subject<Boolean> = BehaviorSubject.createDefault(false)
+    private var mapClicks: Subject<Boolean> = PublishSubject.create()
 
     companion object {
         const val OBJECT_DETAILS = "object-details"
+        const val ZOOM_LEVEL_ONE = 18.0f
+        const val ZOOM_LEVEL_TWO = 19.0f
+        const val ZOOM_LEVEL_THREE = 20.0f
```
[the remark this answers is not in the document]
Comment (sam33rdhakal) [reply map/src/main/kotlin/edu/artic/map/MapFragment.kt]: Totally, agree.

File: base/src/main/java/edu/artic/base/utils/ActivityExtensions.kt [hunk 11]
```diff
@@ -73,3 +73,13 @@ fun Context.isResourceConstrained() : Boolean {
 
     return retVal
 }
+
+
+fun Activity.getStatusBarHeight(): Int {
```
[the remark this answers is not in the document]
Comment (sam33rdhakal) [reply base/src/main/java/edu/artic/base/utils/ActivityExtensions.kt]: Done.

File: map/src/main/kotlin/edu/artic/map/MapActivity.kt [hunk 12]
```diff
@@ -25,6 +37,27 @@ class MapActivity : BaseActivity() {
...
             setOnNavigationItemSelectedListener(NavigationSelectListener(this.context))
         }
+        graphBuilder(tour)
     }
 
+    /**
+     * Passing argument directly to startDestination is not possible without
+     * building graph and setting the bundle manually.
+     */
+    private fun graphBuilder(tour: ArticTour?) {
```
[the remark this answers is not in the document]
Comment (sam33rdhakal) [reply map/src/main/kotlin/edu/artic/map/MapActivity.kt]: Updated.

File: map/src/main/kotlin/edu/artic/map/MapActivity.kt [hunk 13]
```diff
@@ -25,6 +37,27 @@ class MapActivity : BaseActivity() {
...
+        val navHostFragment = container as NavHostFragment
+        val inflater = navHostFragment.navController.navInflater
+        val graph = inflater.inflate(R.navigation.map_navigation_graph)
+
+        val bundle = Bundle().apply {
+            tour?.let {
+                /** using the argument key define in [R.navigation.map_navigation_graph] **/
```
[the remark this answers is not in the document]
Comment (sam33rdhakal) [reply map/src/main/kotlin/edu/artic/map/MapActivity.kt]: This key is for the graph and the Argument's key should be the same as the one defined in the navigation graph.

File: map/src/main/kotlin/edu/artic/map/MapFragment.kt [hunk 14]
```diff
@@ -98,80 +112,10 @@ class MapFragment : BaseViewModelFragment<MapViewModel>() {
...
-                            "    \"featureType\": \"transit\",\n" +
-                            "    \"stylers\": [\n" +
-                            "      {\n" +
-                            "        \"visibility\": \"off\"\n" +
-                            "      }\n" +
-                            "    ]\n" +
-                            "  }\n" +
-                            "]")
-            )
+            map.setMapStyle(MapStyleOptions(requireActivity().assets.fileAsString(filename = "google_map_config.json")))
```
[the remark this answers is not in the document]
Comment (sam33rdhakal) [reply map/src/main/kotlin/edu/artic/map/MapFragment.kt]: Updated.

**Comment by Cliabhach** [comment]
I think this must be the last really large merge. From now on, smaller pieces.

**Pull request merged**